### ISSUE TITLE
Support C# 10 interpolated string builders on TextWriter

### ISSUE THREAD

**Issue by scalablecory** [dotnet/runtime#51573]
#50635 adds an interpolated string builder for `StringBuilder`. We should consider adding a builder for `TextWriter` as well.

**Comment by scalablecory** [comment]
CC @stephentoub

**Comment by tannergooding** [comment]
CC. @stephentoub, is this something you think we should look more into providing?

**Comment by stephentoub** [comment]
> is this something you think we should look more into providing?

It's reasonable to investigate.  I expect a main issue here that could be prohibitive is around the virtual nature of what such overloads would be.  If they're not virtual, then the best you could probably do would be to have a synchronous overload that generates a temporary span/memory that's then written.  If it is virtual, it's not clear how the same handler that would be in the base signature could be customized for derived types, e.g. what would the base virtual method take that would enable an override on StreamWriter to efficiently handle the input (by e.g. formatting directly into its buffer).  It might be we'd instead need to only have methods exposed on the concrete public derived writer types. Regardless, it would need a thought-through design to make forward progress.

**Comment by alex-jitbit** [comment]
This would be so awesome. When ASP.NET `HtmlHelpder`s render some html content, this would allow us write the content more efficiently _and_ keep the code readable.

**Comment by pawchen** [comment]
Yes, the fact that the compiler helps breaking an interpolated string `Write` call into separate `AppendXXX` calls (that forwards to `Write`) avoids the unnecessary `StringBuilder.ToString()` that gets immediately copied then discarded afterwards.

I actually wrote some extension methods to do that in my own project. Unfortunately overload resolution not picking them up, so I need to use names other than `Write` or `WriteLine`.

```C#
internal static class TextWriterExtensions
{
    public static void Inline(this TextWriter writer, [InterpolatedStringHandlerArgument(nameof(writer))] DirectWriteInterpolatedStringHandler handler)
    {
    }

    public static void Line(this TextWriter writer, [InterpolatedStringHandlerArgument(nameof(writer))] DirectWriteInterpolatedStringHandler handler)
    {
        writer.WriteLine();
    }

    [InterpolatedStringHandler]
    public readonly struct DirectWriteInterpolatedStringHandler
    {
        private readonly TextWriter Writer;

        public DirectWriteInterpolatedStringHandler(int literalLength, int formattedCount, TextWriter writer)
        {
            Writer = writer;
        }

        public void AppendLiteral(string? s) => Writer.Write(s);

        public void AppendFormatted(string? s) => Writer.Write(s);

        public void AppendFormatted(ReadOnlySpan<char> s) => Writer.Write(s);

        public void AppendFormatted(int s) => Writer.Write(s);

        public void AppendFormatted<T>(T t) => Writer.Write(t);
    }

    public static void Inline(this TextWriter writer, string? s) => writer.Write(s);

    public static void Line(this TextWriter writer) => writer.WriteLine();

    public static void Line(this TextWriter writer, string? s) => writer.WriteLine(s);
}
```

**Comment by tannergooding** [comment]
Closing as this is inactionable as is due to not having an API proposal.

If we'd like to drive it forward, someone needs to open a new issue following the intended proposal template and covering the actual API surface to be added.

-- While the premise here is sound and the theoretical API surface could be guessed at by the area owners, the backlog is simply to big for us to do all the work on every suggestion. So an interested party needs to drive the base design work here and ensure its in the right format.
